src/vaev-layout/flex.cpp

```diff
@@ -1,102 +1,484 @@
+                // TODO: ignoring breaks for now
+                if (currLineSize + flexItems[endItemIdx].getHypotheticalMainSize() <= availableMainSpace)
+                    currLineSize += flexItems[endItemIdx++].getHypotheticalMainSize();
+                else
+                    break;
+            }
 
-    // TODO: Implement this step
+            flexLines.pushBack(FlexLine{mutSub(flexItems, startItemIdx, endItemIdx)});
+
+            startItemIdx = endItemIdx;
+        }
+    }
 
     // 6. Resolve the flexible lengths
-    // TODO: Implement this step
+    for (auto &flexLine : flexLines) {
+        Px sumItemsHypotheticalMainSizes{0};
+        for (auto const &flexItem : flexLine.items) {
+            sumItemsHypotheticalMainSizes += flexItem.getHypotheticalMainSize();
+        }
+
+        bool matchedSize = sumItemsHypotheticalMainSizes == mainSize;
+        bool flexCaseIsGrow = sumItemsHypotheticalMainSizes < mainSize;
+
+        Vec<FlexItem *> unfrozenItems{flexLine.items.len()}, frozenItems{flexLine.items.len()};
+        Px sumFrozenSizes{0};
+        for (auto &flexItem : flexLine.items) {
+            if (
+                matchedSize or
+                (flexCaseIsGrow and flexItem.flexBaseSize > flexItem.getHypotheticalMainSize()) or
+                (!flexCaseIsGrow and flexItem.flexBaseSize < flexItem.getHypotheticalMainSize()) or
+                (flexItem.frag.style->flex->grow == 0 and flexItem.frag.style->flex->shrink == 0)
+            ) {
+                flexItem.usedMainSize = flexItem.getHypotheticalMainSize() + flexItem.output.margins.horizontal();
+                frozenItems.pushBack(&flexItem);
+                sumFrozenSizes += flexItem.usedMainSize;
+            } else {
+                flexItem.usedMainSize = flexItem.getHypotheticalMainSize() + flexItem.output.margins.horizontal();
+                unfrozenItems.pushBack(&flexItem);
+            }
+        }
+
+        auto computeStats = [&]() {
+            Px sumOfUnfrozenSizes{0};
+            Number sumUnfrozenFlexFactors{0};
+            for (auto *flexItem : unfrozenItems) {
+                sumOfUnfrozenSizes += flexItem->usedMainSize;
+                sumUnfrozenFlexFactors += flexCaseIsGrow ? flexItem->frag.style->flex->grow : flexItem->frag.style->flex->shrink;
+            }
+
+            return Tuple<Px, Number>(sumOfUnfrozenSizes, sumUnfrozenFlexFactors);
+        };
+
+        auto [sumUnfrozenSizes, _] = computeStats();
+        // FIXME: weird types of spaces and sizes here, since free space can be negative
+        Number initialFreeSpace = Number{mainSize} - Number{sumUnfrozenSizes + sumFrozenSizes};
+
+        while (unfrozenItems.len()) {
+            auto [sumUnfrozenSizes, sumUnfrozenFlexFactors] = computeStats();
+            auto freeSpace = Number{mainSize} - Number{sumUnfrozenSizes + sumFrozenSizes};
+
+            if (sumUnfrozenFlexFactors < 1 && abs(initialFreeSpace * sumUnfrozenFlexFactors) < abs(freeSpace))
```

Comment:
Here, you can use Math::abs() on fixed number types

Reply:
lets leave this open for discussion tmrw